File: GFG POTD/019-June-2025/006-search-pattern-rabin-karp-algorithm.cpp
// https://www.geeksforgeeks.org/problems/search-pattern-rabin-karp-algorithm--141631/1




#include <bits/stdc++.h>
using namespace std;

class Solution {
  private:
    typedef long long ll;
    const int M = 1e9 + 7;
    int m, n;
    
    ll getHash(string &pat) {
        ll ans = 0, factor = 1;
        for (int i = m - 1; i >= 0; i --) {
            ans   += (factor * (pat[i] - 'a')) % M;
            factor = (factor * 26) % M;  // 26^m, 26^m-1, 26^m-2, ..., 26^0
        }
        return ans % M;
    }
  public:
    // TC : O(M + N)
    // SC : O(1)
    vector<int> search(string &pat, string &txt) {
        m = pat.length(), n = txt.length();
        
        ll dPowerM = 1;
        for (int i = 1; i <= m; i ++) {
            dPowerM = (dPowerM * 26) % M;
        }
        
        ll hashPatt = getHash(pat);
        ll hashTxt  = 0;
        vector<int> result;
        
        for (int i = 0; i <= n - m; i ++) {

            if (i == 0) {
                string s = txt.substr(i, m);
                hashTxt = getHash(s);
            } else {    // calculate rolling hash
                hashTxt = (hashTxt * 26 - (dPowerM * (txt[i - 1] - 'a')) + (txt[i + m - 1] - 'a')) % M;
                hashTxt = (hashTxt + M) % M;
            }
            
            if (hashPatt == hashTxt) {
                int j = 0;
                for (; j < m; j ++) {
                    if (txt[i + j] != pat[j]) {
                        break;
                    }
                }
                if (j == m) {
                    result.push_back(i + 1);
                }
            }
        }
        
        return result;
    }
};

File: GFG POTD/019-June-2025/006-search-pattern-rabin-karp-algorithm.java
// https://www.geeksforgeeks.org/problems/search-pattern-rabin-karp-algorithm--141631/1




import java.util.*;

class Solution {
    private static final int M = 1_000_000_007;
    
    // TC : O(M + N)
    // SC : O(1)
    ArrayList<Integer> search(String pat, String txt) {
        int m = pat.length(), n = txt.length();
        ArrayList<Integer> result = new ArrayList<>();
        
        long powerM = 1;
        for (int i = 1; i <= m; i ++) {
            powerM = (powerM * 26) % M;
        }
        
        long hashPat = getHash(pat);
        long hashTxt = 0;
        
        for (int i = 0; i <= n - m; i ++) {
            if (i == 0) {
                String str = txt.substring(i, i + m);
                hashTxt = getHash(str);
            } else {
                hashTxt = (hashTxt * 26 - (powerM * (txt.charAt(i - 1) - 'a')) + (txt.charAt(i + m - 1) - 'a')) % M;
                hashTxt = (hashTxt + M) % M;
            }
            
            if (hashTxt == hashPat) {
                int j = 0;
                for (; j < m; j ++) {
                    if (txt.charAt(i + j) != pat.charAt(j)) break;
                }
                if (j == m) {
                    result.add(i + 1);
                }
            }
        }
        
        return result;
    }
    
    private long getHash(String s) {
        long ans = 0, factor = 1;
        for (int i = s.length() - 1; i >= 0; i --) {
            ans    = (ans + factor * (s.charAt(i) - 'a')) % M;
            factor = (factor * 26) % M;
        }
        return ans % M;
    }
}